File: py/medium/0451/0451.py
from collections import Counter
from operator import itemgetter


class Solution:
    def frequencySort(self, s: str) -> str:
        sorted_list = sorted(Counter(s).items(), key=itemgetter(1), reverse=True)
        result = [item[0] * item[1] for item in sorted_list]
        return "".join(result)


solution = Solution()
assert solution.frequencySort("tree") == "eetr"
assert solution.frequencySort("cccaaa") == "cccaaa"
assert solution.frequencySort("Aabb") == "bbAa"

File: elixir/medium/0451/0451.exs
defmodule Solution do
  @spec frequency_sort(s :: String.t()) :: String.t()
  def frequency_sort(s) do
    s
    |> String.graphemes()
    |> Enum.frequencies()
    |> Enum.sort(fn {_, v1}, {_, v2} -> v1 > v2 end)
    |> Enum.map(fn {k, v} -> String.duplicate(k, v) end)
    |> Enum.join("")
  end
end

ExUnit.start()

defmodule AssertionTest do
  use ExUnit.Case, async: true

  test "frequency_sort 1" do
    assert Solution.frequency_sort("tree") == "eetr"
  end

  test "frequency_sort 2" do
    assert Solution.frequency_sort("cccaaa") == "cccaaa"
  end

  test "frequency_sort 3" do
    assert Solution.frequency_sort("bbaA") == "bbaA"
  end
end
